modules/lldp/announce.c

```diff
@@ -0,0 +1,191 @@
+	uint16_t ttl_value = rte_cpu_to_be_16(lldp_ctx.ttl);
+	struct rte_ether_addr src_addr;
+	struct nexthop *local_ip;
+
+	iface_get_eth_addr(iface->id, &src_addr);
+	rte_pktmbuf_reset(mbuf);
+
+	lldp_tlv_append(
+		mbuf, T_CHASSIS_ID, T_CHASSIS_MAC_ADDRESS, RTE_ETHER_ADDR_LEN, src_addr.addr_bytes
+	);
+
+	lldp_tlv_append(mbuf, T_CHASSIS_ID, T_CHASSIS_IF_ALIAS, strlen(SYSNAME), SYSNAME);
+	lldp_tlv_append(
+		mbuf, T_PORT_ID, T_PORT_MAC_ADDRESS, RTE_ETHER_ADDR_LEN, src_addr.addr_bytes
+	);
+	lldp_tlv_append(mbuf, T_PORT_ID, T_PORT_IF_ALIAS, strlen(iface->name), iface->name);
+	lldp_tlv_append(mbuf, T_TTL, T_NO_SUBTYPE, sizeof(ttl_value), &ttl_value);
+	lldp_tlv_append(mbuf, T_PORT_DESC, T_NO_SUBTYPE, strlen(iface->name), iface->name);
+	lldp_tlv_append(mbuf, T_SYSTEM_NAME, T_NO_SUBTYPE, strlen(SYSNAME), SYSNAME);
+	lldp_tlv_append(mbuf, T_SYSTEM_DESC, T_NO_SUBTYPE, strlen(SYSDESC), SYSDESC);
+
+	local_ip = ip4_addr_get(iface->id);
+	if (local_ip) {
+		lldp_ip4.ip4_addr = local_ip->ip;
+		lldp_tlv_append(
+			mbuf,
+			T_CHASSIS_ID,
+			T_CHASSIS_NET_ADDRESS,
+			sizeof(struct lldp_ip4),
+			&lldp_ip4
+		);
+	}
+	// TOOD: ip6_addr_get(iface->id)
+
+	return lldp_tlv_append(mbuf, T_END, T_NO_SUBTYPE, 0, NULL) != NULL;
+}
+
+uint64_t last_sent = 0;
+
+static uint16_t
+lldp_announce_process(struct rte_graph *graph, struct rte_node *node, void **, uint16_t) {
+	struct eth_output_mbuf_data *eth_data;
+	const struct iface *iface;
+	struct rte_mbuf *mbuf;
+	uint16_t sent;
+	uint64_t now;
+	uint16_t n;
+
+	sent = 0;
+
+	now = rte_get_tsc_cycles();
+	n = rte_ring_dequeue_burst(lldp_ring, node->objs, RTE_GRAPH_BURST_SIZE, NULL);
+
+	for (unsigned i = 0; i < n; i++) {
+		iface = node->objs[i];
+		mbuf = rte_pktmbuf_alloc(lldp_pool);
+
+		if (build_lldp_frame(mbuf, iface) == false) {
+			rte_node_enqueue_x1(graph, node, ERROR, mbuf);
+		}
+
+		// Prepare ethernet layer info.
+		eth_data = eth_output_mbuf_data(mbuf);
+		rte_ether_addr_copy(&lldp_dst, &eth_data->dst);
+		eth_data->ether_type = rte_cpu_to_be_16(RTE_ETHER_TYPE_LLDP);
+		eth_data->iface = iface;
+
+		rte_node_enqueue_x1(graph, node, OUTPUT, mbuf);
+		sent++;
+	}
+	last_sent = now;
+
+	return sent;
+}
+
+static void lldp_announce_register(void) {
+	lldp_ring = rte_ring_create(
+		"lldp_announce",
+		RTE_GRAPH_BURST_SIZE,
+		SOCKET_ID_ANY,
+		RING_F_MP_RTS_ENQ | RING_F_MC_RTS_DEQ
+	);
+	if (lldp_ring == NULL)
+		ABORT("rte_ring_create(arp_output_request): %s", rte_strerror(rte_errno));
+
+	lldp_pool = rte_pktmbuf_pool_create(
+		"lldp_announce",
+		RTE_GRAPH_BURST_SIZE,
+		64, // cache_size
+		0, // priv_size
+		RTE_MBUF_DEFAULT_BUF_SIZE,
+		SOCKET_ID_ANY
+	);
+	if (lldp_pool == NULL)
+		ABORT("rte_pktmbuf_pool_create(lldp_announce): %s", rte_strerror(rte_errno));
+}
+
+static void lldp_announce_unregister(void) {
+	rte_ring_free(lldp_ring);
+	rte_mempool_free(lldp_pool);
+}
+
+static struct rte_node_register lldp_announce_node = {
+	.flags = RTE_NODE_SOURCE_F,
+	.name = "lldp_announce",
```

Comment:
Same remark, I think it would be more consistent with the rest of the nodes if it was called lldp_output.

Reply:
I chose `lldp_announce` as it is a source node, it felt weird to have an `*_output` node as a source node.
WDYT ?

Reply:
I agree that it feels weird but it would be consistent with arp_output_request. Keep in mind that node names show up in the software stats. Users don't see that it is a source node unless they dump the graph and convert it to an image.